File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        string currentLargest = "";
        string newSubString = "";
        string currentSubString = "";
        bool isMatched;

        for (int i = 0; i < s.length(); i++) {
            isMatched = false;
            for (int j = 0; j < currentSubString.length(); j++) {
                if (s[i] == currentSubString[j]) {
                    if (currentLargest.length() < currentSubString.length()) {
                        currentLargest = currentSubString;
                    }
                    newSubString = "";
                    isMatched = true;
                    continue;
                }
                if (isMatched)
                    newSubString += currentSubString[j];
            }
            if (isMatched)
                currentSubString = newSubString;
            currentSubString += s[i];
        }
        if (currentLargest.length() < currentSubString.length()) {
            currentLargest = currentSubString;
        }

        return currentLargest.length();
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        Dictionary<char, int> map = new Dictionary<char,int>();
        int maxLength = 0;

        int left =0;
        int right = 0;

        while(right < s.Length){
            if (!map.ContainsKey(s[right]))
    map[s[right]] = 0;

map[s[right]]++;


            if(map[s[right]] > 1){
                // shrink
                while(map[s[right]] != 1){
                    map[s[left]] -= 1;
                    left+=1;
                }
            }
            if(right - left + 1 > maxLength)
                maxLength = right - left + 1;

            right += 1;
        }

        return maxLength;
    }
}